### PR TITLE
pkg/pkg.mk: Avoid doing a git fetch if we already have the commit.

### PR DESCRIPTION
**Pull request by jcarrano**
### Contribution description

D'oh! I screwed up a little in #11129 and a git fetch was being done for each package every time the user ran make.

If the commit is already in the local repo, we should not need to do any network operation.

### Testing procedure

Type `make` twice in any example that uses packages. See how it is fetched twice. With this patch it is not.

### Issues/PRs references

Bug introduced in #11129.

### CODE REVIEW

File: pkg/pkg.mk [hunk 1]
```diff
@@ -32,7 +32,9 @@ endif
 git-ensure-version: $(PKG_BUILDDIR)/.git-downloaded
 	if [ $(shell git -C $(PKG_BUILDDIR) rev-parse HEAD) != $(PKG_VERSION) ] ; then \
 		git -C $(PKG_BUILDDIR) clean -xdff ; \
-		git -C $(PKG_BUILDDIR) fetch "$(PKG_URL)" "$(PKG_VERSION)" ; \
+		if git cat-file -e -t '$(PKG_VERSION)^{commit}' ; then \
```
Comment (cladmi) [inline pkg/pkg.mk]: On `ubuntu` bionic, git does not like `-e -t`.
`error: switch `t' : incompatible with -e`.

I would like to maybe put it outside in a "git_check_version_is_a_commit" so document that it detects branches.

Comment (cladmi) [reply pkg/pkg.mk]:
```
git cat-file -e 2019.04^{commit} && echo commit
commit
```
```
git cat-file -e 2019.04-branch^{commit}
fatal: Not a valid object name 2019.04-branch^{commit}
```

Also works with a hash as expected.

**Comment by cladmi** [comment]
It does not re-fetch with `tests/pkg_c25519` in master already.

**Comment by kaspar030** [comment]
> t does not re-fetch with `tests/pkg_c25519` in master already.

c25519 is not using git.

**Comment by cladmi** [comment]
> > t does not re-fetch with `tests/pkg_c25519` in master already.
> 
> c25519 is not using git.

Ahhhh did not pay attention to that ><

**Comment by kaspar030** [comment]
This PR is an improvement (unnecessary fetches are avoided). ACK.

**Comment by kaspar030** [comment]
please rebase & squash!

**Comment by kaspar030** [comment]
There are some other issues with "git_ensure_version":

1. if the package specifies a tag (instead of commit hash), the initial comparison always fails.

2. if the package is using patches, the initial comparison always fails.

The following "git clean" removes ".git_downloaded" and ".git_patched" unconditionally, causing any recursive make to re-do the clean and patch again.

**Comment by kaspar030** [comment]
I found another issue: this currently does "git -Cpkdir clean -xdff". If, for any reason, the pkgdir is *not* a git folder, git climbs up all the way to the RIOT checkout (looking for the git folder), **then cleans that**.

A check whether the to-be-cleaned folder is a git checkout needs to be added.

@jcarrano @MrKevinWeiss I think this is release critical, I lost quite some work because of this.

**Comment by MrKevinWeiss** [comment]
>  I think this is release critical, I lost quite some work because of this.

Noted, I pull push for it

**Comment by cladmi** [comment]
> I found another issue: this currently does `git -C<pkgdir> clean -xdff`. If, for any reason, the pkgdir is _not_ a git folder, git climbs up all the way to the RIOT checkout (looking for the git folder), **then cleans that**.

All the code from `pkg.mk` supposes it is a git repository. The `clean` target does the same. Why would it not be one ?

**Comment by cladmi** [comment]
For security, the `git` command could be replaced by `git --git-dir=.git -C $(PKGDIR)` as `PKGDIR` is supposed to be the `root` of the repository here.

However, this is somehow unrelated

**Comment by MrKevinWeiss** [comment]
@jcarrano Is there anything holding up the squash?

**Comment by fjmolinas** [comment]
I can't reproduce this issue in master, I believe this is fixed, can I close?

**Comment by kaspar030** [comment]
> I can't reproduce this issue in master, I believe this is fixed, can I close?

nope, it is just quieted. Use e.g., this patch: https://gist.github.com/kaspar030/d0bcf4f38d91d2b7e785b8327da608be/raw

... then type make a couple of times in a git package with patches, e.g., `tests/pkg_heatshrink`.

**Comment by fjmolinas** [comment]
> nope, it is just quieted. Use e.g., this patch: https://gist.github.com/kaspar030/d0bcf4f38d91d2b7e785b8327da608be/raw

That is not master.

**Comment by fjmolinas** [comment]
- First build:

```
make: Entering directory '/home/francisco/workspace/RIOT/tests/pkg_heatshrink'
Building application "tests_pkg_heatshrink" for "native" with MCU "native".

[INFO] cloning heatshrink
Cloning into '/home/francisco/workspace/RIOT/tests/pkg_heatshrink/bin/pkg/native/heatshrink'...
remote: Enumerating objects: 442, done.
remote: Total 442 (delta 0), reused 0 (delta 0), pack-reused 442
Receiving objects: 100% (442/442), 177.36 KiB | 868.00 KiB/s, done.
Resolving deltas: 100% (264/264), done.
HEAD is now at 7d419e1 Update contact info.
[INFO] updating heatshrink /home/francisco/workspace/RIOT/tests/pkg_heatshrink/bin/pkg/native/heatshrink/.pkg-state.git-downloaded
echo 7d419e1fa4830d0b919b9b6a91fe2fb786cf3280 > /home/francisco/workspace/RIOT/tests/pkg_heatshrink/bin/pkg/native/heatshrink/.pkg-state.git-downloaded
[INFO] patch heatshrink
"make" -C /home/francisco/workspace/RIOT/pkg/heatshrink
"make" -C /home/francisco/workspace/RIOT/tests/pkg_heatshrink/bin/pkg/native/heatshrink -f /home/francisco/workspace/RIOT/pkg/heatshrink/Makefile.heatshrink
"make" -C /home/francisco/workspace/RIOT/boards/native
"make" -C /home/francisco/workspace/RIOT/boards/native/drivers
"make" -C /home/francisco/workspace/RIOT/core
"make" -C /home/francisco/workspace/RIOT/cpu/native
"make" -C /home/francisco/workspace/RIOT/cpu/native/periph
"make" -C /home/francisco/workspace/RIOT/cpu/native/stdio_native
"make" -C /home/francisco/workspace/RIOT/cpu/native/vfs
"make" -C /home/francisco/workspace/RIOT/drivers
"make" -C /home/francisco/workspace/RIOT/drivers/periph_common
"make" -C /home/francisco/workspace/RIOT/sys
"make" -C /home/francisco/workspace/RIOT/sys/auto_init
"make" -C /home/francisco/workspace/RIOT/sys/embunit
"make" -C /home/francisco/workspace/RIOT/sys/test_utils/interactive_sync
   text    data     bss     dec     hex filename
  39465     668   50088   90221   1606d /home/francisco/workspace/RIOT/tests/pkg_heatshrink/bin/native/tests_pkg_heatshrink.elf
make: Leaving directory '/home/francisco/workspace/RIOT/tests/pkg_heatshrink'
```

- second:

```
make: Entering directory '/home/francisco/workspace/RIOT/tests/pkg_heatshrink'
Building application "tests_pkg_heatshrink" for "native" with MCU "native".

make[1]: Nothing to be done for 'prepare'.
"make" -C /home/francisco/workspace/RIOT/pkg/heatshrink
"make" -C /home/francisco/workspace/RIOT/tests/pkg_heatshrink/bin/pkg/native/heatshrink -f /home/francisco/workspace/RIOT/pkg/heatshrink/Makefile.heatshrink
"make" -C /home/francisco/workspace/RIOT/boards/native
"make" -C /home/francisco/workspace/RIOT/boards/native/drivers
"make" -C /home/francisco/workspace/RIOT/core
"make" -C /home/francisco/workspace/RIOT/cpu/native
"make" -C /home/francisco/workspace/RIOT/cpu/native/periph
"make" -C /home/francisco/workspace/RIOT/cpu/native/stdio_native
"make" -C /home/francisco/workspace/RIOT/cpu/native/vfs
"make" -C /home/francisco/workspace/RIOT/drivers
"make" -C /home/francisco/workspace/RIOT/drivers/periph_common
"make" -C /home/francisco/workspace/RIOT/sys
"make" -C /home/francisco/workspace/RIOT/sys/auto_init
"make" -C /home/francisco/workspace/RIOT/sys/embunit
"make" -C /home/francisco/workspace/RIOT/sys/test_utils/interactive_sync
   text    data     bss     dec     hex filename
  39465     668   50088   90221   1606d /home/francisco/workspace/RIOT/tests/pkg_heatshrink/bin/native/tests_pkg_heatshrink.elf
make: Leaving directory '/home/francisco/workspace/RIOT/tests/pkg_heatshrink'
```

**Comment by fjmolinas** [comment]
> > nope, it is just quieted. Use e.g., this patch: https://gist.github.com/kaspar030/d0bcf4f38d91d2b7e785b8327da608be/raw
> 
> That is not master.

Or maybe this one is still a smarter way of fetching event if its not fetch on each make incocation?

**Comment by aabadie** [comment]
The problem targeted by this PR was fixed by #13036, which is already merged. I'm closing this one but feel free to re-open if you disagree.

**Comment by fjmolinas** [comment]
I don't thing it is fixed #13036 does not address smart fetching.

**Comment by aabadie** [comment]
> I don't thing it is fixed #13036 does not address smart fetching.

Can you elaborate on this ? I'm trying to reproduce the issue on master and I can't: git fetch is not called anymore on a package that is already cloned/patched.

**Comment by maribu** [comment]
@fjmolinas: Ping?

**Comment by kaspar030** [comment]
Is this still a thing? Anyhow, I'll postpone this as it doesn't seem to break anything ATM.

**Pull request closed without merging**